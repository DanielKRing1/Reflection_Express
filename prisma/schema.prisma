// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_LOCAL")
}

model User {
  password_userId             Password?           @relation(name: "password_userId")
  journal_userId              Journal[]           @relation(name: "journal_userId")
  email                       String              @id @unique

  // Set to null when delete Journal + Update id if Journal id updated
  user_lastUsedJournalId      Journal?            @relation(name: "user_lastUsedJournalId", fields: [lastUsedJId], references: [id], onDelete: SetNull, onUpdate: Cascade)
  lastUsedJId                 BigInt?             @unique
}

model Password {
  // Cascade delete/update with User
  password_userId             User                @relation(name: "password_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String              @id

  hash                        String
}

model Journal {
  user_lastUsedJournalId      User?               @relation(name: "user_lastUsedJournalId")
  inkling_journalId           Inkling[]           @relation(name: "inkling_journalId")
  journalentry_journalId      JournalEntry[]      @relation(name: "journalentry_journalId")
  thought_journalId           Thought[]           @relation(name: "thought_journalId")
  reflection_journalId        Reflection[]        @relation(name: "reflection_journalId")
  id                          BigInt              @id @default(autoincrement())
  
  // Cascade delete/update with User
  journal_userId              User                @relation(name: "journal_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String

  @@index([userId])
}

model Inkling {
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  inkling_journalId           Journal             @relation(name: "inkling_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model JournalEntry {
  reflection_journalEntryId   Reflection[]        @relation(name: "reflection_journalEntryId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  journalentry_journalId      Journal             @relation(name: "journalentry_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  @@id([journalId, timeId])
}

model Thought {
  reflection_thoughtId        Reflection[]        @relation(name: "reflection_thoughtId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  thought_journalId           Journal             @relation(name: "thought_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model Reflection {
  // Cascade delete/update with Journal
  reflection_journalId        Journal             @relation(name: "reflection_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  // Cascade delete/update with JournalEntry
  reflection_journalEntryId   JournalEntry        @relation(name: "reflection_journalEntryId", fields: [journalId, journalEntryId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  journalEntryId              DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Thought
  reflection_thoughtId        Thought             @relation(name: "reflection_thoughtId", fields: [journalId, thoughtId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  thoughtId                   DateTime            @db.Timestamptz(3)

  decision                    Int                 @db.SmallInt

  @@id([journalEntryId, thoughtId])
}
