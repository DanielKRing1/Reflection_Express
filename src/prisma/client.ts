import { PrismaClient } from "./generated/client";

let prisma = new PrismaClient({
    // log: ["query"],
});

export default prisma;
