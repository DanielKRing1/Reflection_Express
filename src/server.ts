// EXPRESS
import express, { Express } from "express";
import cors from "cors";

// MIDDLEWARE
import loggingMiddleware from "./middlewares/logging.middleware";

// ROUTES
import genRouter from "./routes";

// CONFIG
import { PORT } from "./config/server.config";

// APP
const app: Express = express();

// MIDDLEWARE
app.use(cors());
// parse requests of content-type - application/json
app.use(express.json());
// parse requests of content-type - application/x-www-form-urlencoded
app.use(
  express.urlencoded({
    extended: true,
  })
);
app.use(loggingMiddleware);

// GRAPHQL
app.use(genRouter());

// START SERVER
app.listen(PORT, () => console.log(`Server listening on port ${PORT}!`));
