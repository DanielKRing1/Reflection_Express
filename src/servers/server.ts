import genExpressApp from "./app";
import gqlServer from "./gqlServer";

import accessPromise from "../middlewares/session/access";

export default (async () => {
    const app = await genExpressApp();
    const access = await accessPromise();

    return gqlServer(app, access);
})();
