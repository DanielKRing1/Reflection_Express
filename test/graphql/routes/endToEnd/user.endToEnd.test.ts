import http from "http";
import supertest from "supertest";

import setCookie from "set-cookie-parser";

import server from "../../../../src/index";
import prisma from "../../../../src/prisma/client";
import {
    ACCESS_SESSION_COOKIE_NAME,
    META_ACCESS_SESSION_COOKIE_NAME,
} from "../../../../src/middlewares/session/access/constants";
import {
    REFRESH_SESSION_COOKIE_NAME,
    META_REFRESH_SESSION_COOKIE_NAME,
} from "../../../../src/middlewares/session/refresh/constants";

const USER_ID = "1";
const PASSWORD = "123";

describe("User GraphQL server", () => {
    let s: http.Server;
    let agent: supertest.SuperAgentTest;

    beforeAll(async () => {
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;

        s = await server;
        agent = supertest.agent(s);
    });

    it("Should get access and refresh session cookies (at least 2 cookies)", async () => {
        const res = await agent
            .post("/login/create-user")
            .send({ userId: USER_ID, password: PASSWORD });

        const cookies = setCookie
            .parse(res as any)
            .sort((a, b) => a.name.localeCompare(b.name));
        console.log(cookies);
        expect(cookies.length).toBe(4);

        expect(cookies[0].name).toBe(ACCESS_SESSION_COOKIE_NAME);
        expect(cookies[1].name).toBe(META_ACCESS_SESSION_COOKIE_NAME);
        expect(cookies[2].name).toBe(REFRESH_SESSION_COOKIE_NAME);
        expect(cookies[3].name).toBe(META_REFRESH_SESSION_COOKIE_NAME);
    });

    it("Should get the created User (this requires successfully sending an access session cookie with the request)", async () => {
        const queryData = {
            query: `query User {
                user {
                    email
                    lastUsedJId
                }
            }`,
            variables: { email: USER_ID },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should fail to get the created User without an access cookie", async () => {
        // Reset agent (and cookies)
        agent = supertest.agent(s);

        // Perform test
        const queryData = {
            query: `query User {
                user {
                    email
                    lastUsedJId
                }
            }`,
            variables: { email: USER_ID },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should login and then get the created User", async () => {
        await agent
            .post("/login")
            .send({ userId: USER_ID, password: PASSWORD });

        // Perform test
        const queryData = {
            query: `query User {
                user {
                    email
                    lastUsedJId
                }
            }`,
            variables: { email: USER_ID },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should fail login with wrong password and then fail to get the created User without an access cookie", async () => {
        // Reset agent (and cookies)
        agent = supertest.agent(s);

        // Unauthenticated login
        await agent
            .post("/login")
            .send({ userId: USER_ID, password: "wrong password" });

        // Perform test
        const queryData = {
            query: `query User {
                user {
                    email
                    lastUsedJId
                }
            }`,
            variables: { email: USER_ID },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    it("Should fail login with wrong userId and then fail to get the created User without an access cookie", async () => {
        // Reset agent (and cookies)
        agent = supertest.agent(s);

        // Unauthenticated login
        await agent
            .post("/login")
            .send({ userId: "wrong userId", password: PASSWORD });

        // Perform test
        const queryData = {
            query: `query User {
                user {
                    email
                    lastUsedJId
                }
            }`,
            variables: { email: USER_ID },
        };

        const res = await agent.post("/graphql").send(queryData);

        expect(res.text).toMatchSnapshot();
    });

    afterAll(async () => {
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;

        await prisma.$disconnect();
        await s.close();
    });
});
