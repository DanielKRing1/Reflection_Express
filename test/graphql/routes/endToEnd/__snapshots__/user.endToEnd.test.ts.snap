// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User GraphQL server Should fail login with wrong password and then fail to get the created User without an access cookie 1`] = `
"{"errors":[{"message":"you must be logged in to query this schema","locations":[{"line":2,"column":17}],"path":["user"],"extensions":{"code":"UNAUTHENTICATED"}}],"data":{"user":null}}
"
`;

exports[`User GraphQL server Should fail login with wrong userId and then fail to get the created User without an access cookie 1`] = `
"{"errors":[{"message":"you must be logged in to query this schema","locations":[{"line":2,"column":17}],"path":["user"],"extensions":{"code":"UNAUTHENTICATED"}}],"data":{"user":null}}
"
`;

exports[`User GraphQL server Should fail to get the created User without an access cookie 1`] = `
"{"errors":[{"message":"you must be logged in to query this schema","locations":[{"line":2,"column":17}],"path":["user"],"extensions":{"code":"UNAUTHENTICATED"}}],"data":{"user":null}}
"
`;

exports[`User GraphQL server Should get the created User (this requires successfully sending an access session cookie with the request) 1`] = `
"{"data":{"user":{"email":"1","lastUsedJId":null}}}
"
`;

exports[`User GraphQL server Should login and then get the created User 1`] = `
"{"data":{"user":{"email":"1","lastUsedJId":null}}}
"
`;
